src/odemis/odemisd/main.py

```diff
@@ -321,78 +322,95 @@ def _terminate_all_alive(self):
+                # just pick a random component
                 logging.warning("All the components to terminate have parents: %s",
-                                parents)
-                parent_less = (tuple(parents.keys())[0],)
-
-            for c in parent_less:
-                cname = c.name
-                logging.debug("Stopping comp %s", cname)
-                # TODO: update the .alive VA every time a component is stopped?
-                # maybe not necessary as we are finishing _everything_
-                try:
-                    # FIXME: don't try to terminate the children created by
-                    # delegation after terminating their parent (and container)
-                    # For now it just causes a warning message, but it'd be
-                    # better to not have the warning at all.
-                    c.terminate()
-                except Exception:
-                    logging.warning("Failed to terminate component '%s'", cname, exc_info=True)
-
-                # Terminate the container if that was the component for which it
-                # was created.
-                # TODO: check there is really no component still running in the
-                # container?
-                if cname in self._instantiator.sub_containers:
-                    container = self._instantiator.sub_containers[cname]
-                    logging.debug("Stopping container %s", container)
-                    try:
-                        container.terminate()
-                    except Exception:
-                        logging.warning("Failed to terminate container %r", container, exc_info=True)
-                    del self._instantiator.sub_containers[cname]
-
-                # remove from the graph
-                del parents[c]
-                for p in parents.values():
-                    p.discard(c)
-                alive.discard(c)
-                try:
-                    mic.alive.value = alive
-                except Exception:
-                    logging.warning("Failed to update the alive VA", exc_info=True)
+                                self._dependents)
+                independents = tuple(self._dependents.keys())[:1]
+
+            for comp in independents:
+                terminating_comps.append(comp)
+                f = executor.submit(self._terminate_independent, comp)
+                fs_running.add(f)
+
+            # Block until one future is completed, then continue looping
+            done, fs_running = futures.wait(fs_running, return_when=futures.FIRST_COMPLETED)
+
+        logging.debug("Finished requesting termination of all components. Waiting for %s components to terminate." % len(fs_running))
+        futures.wait(fs_running, return_when=futures.ALL_COMPLETED)
+
+    def _terminate_independent(self, comp):
+        # First terminate the children
+        children = comp.children.value
+        for child in children:
+            self._terminate_component(child)
```

Comment:
should we do this here explicitly or is it better to just let the parent deal with the children

Reply:
I think you could let the parent do it ... but you'd need to make sure to update .alive properly. So it might be easier to leave it as is and just say we explicitly terminate every component. Somehow, this keeps the rule simple, which is nice too.